train.py

```diff
@@ -406,25 +405,40 @@ def train_loop_fn(device, trainer, loader, last_batch_index):
         This is the main training loop. It trains for 1 epoch.
         """
 
-        def print_training_update(trainer, progress, args, i, tracker):
+        def print_training_update(trainer, progress, args, i):
             stats = get_training_stats(trainer, args=args)
-            stats['rate'] = tracker.rate()
             stats['now'] = now()
             progress.log(stats, tag='train', step=trainer.get_num_updates())
             progress.print_mid_epoch(i+1, force=True)
 
         stats, log_output, skip_stat_keys = None, None, {'clip'}
-        tracker = xm.RateTracker()
+        max_update = args.max_update or math.inf
         for i, samples in enumerate(loader, start=epoch_itr.iterations_in_epoch):
             if i == last_batch_index:
                 # last batches are incomplete
                 break
             log_output = trainer.train_step(samples)
-            tracker.add(sum(sample['nsentences'] for sample in samples))
             reset_perf_training_meters(trainer, i, ignore_index=10)
             if (not (i % args.log_steps)) or (i == last_batch_index-1):
-                step_args = trainer, progress, args, i, tracker
+                step_args = trainer, progress, args, i
                 xm.add_step_closure(print_training_update, args=step_args)
+            num_updates = trainer.get_num_updates()
+            if (
```

Comment:
Is this your syntax to break down statement, or it's theirs?

Reply:
this is their syntax, see [upstream](https://github.com/pytorch/fairseq/blob/master/fairseq/trainer.py#L92-L96) code for example.